File: apps/python/src/exercises/web/controller_logging.py
"""Structured controller logging (see .cursor/skills/controller-logging/)."""

from __future__ import annotations

import logging
from typing import Any

from exercises.web.correlation import CorrelationFilter, current_correlation, merge_correlation
from exercises.web.request_snapshot import request_body, request_headers, request_url_params

_RECORD_STANDARD = frozenset(logging.makeLogRecord({}).__dict__.keys()) | frozenset(
    {"message", "asctime"}
)


def http_request_fields() -> dict[str, Any]:
    """Headers, query params, and JSON/form body for the current Flask request."""
    try:
        from flask import request as flask_request
    except RuntimeError:
        return {}
    return {
        "headers": request_headers(flask_request),
        "url_params": request_url_params(flask_request),
        "body": request_body(flask_request),
    }


def log_received(
    logger: logging.Logger,
    handler: str,
    source: str,
    method: str,
    path: str,
    *,
    include_http_request: bool = True,
    **params: Any,
) -> None:
    fields = dict(params)
    if include_http_request:
        for key, value in http_request_fields().items():
            fields.setdefault(key, value)
    logger.info(
        f"{handler} request received",
        extra=_extra(source, handler, method=method, path=path, **fields),
    )


def log_succeeded(
    logger: logging.Logger,
    handler: str,
    source: str,
    **params: Any,
) -> None:
    logger.info(
        f"{handler} succeeded",
        extra=_extra(source, handler, **params),
    )


def log_warn(
    logger: logging.Logger,
    handler: str,
    source: str,
    message: str,
    **params: Any,
) -> None:
    logger.warning(
        message,
        extra=_extra(source, handler, **params),
    )


def log_error(
    logger: logging.Logger,
    handler: str,
    source: str,
    message: str,
    *,
    exc: BaseException | None = None,
    **params: Any,
) -> None:
    logger.error(
        message,
        extra=_extra(source, handler, **params),
        exc_info=exc is not None,
    )


def log_trace(
    logger: logging.Logger,
    handler: str,
    source: str,
    message: str,
    **params: Any,
) -> None:
    logger.log(
        logging.DEBUG,
        message,
        extra=_extra(source, handler, **params),
    )


def log_kafka_received(
    logger: logging.Logger,
    handler: str,
    source: str,
    topic: str,
    *,
    request_id: str | None = None,
    **params: Any,
) -> None:
    """Kafka/async handler entry (no HTTP access log); may include request_id for correlation."""
    fields = dict(params)
    if request_id:
        fields.setdefault("request_id", request_id)
    logger.info(
        f"{handler} kafka message received",
        extra=_extra(
            source,
            handler,
            method="KAFKA",
            path=f"kafka/{topic}",
            topic=topic,
            **fields,
        ),
    )


def http_access_session_fields() -> dict[str, str]:
    """Top-level correlation for http.request access logs."""
    return current_correlation()


def _extra(source: str, handler: str, **fields: Any) -> dict[str, Any]:
    extra: dict[str, Any] = {"source": source, "controller": handler}
    extra = merge_correlation(extra)
    for key, value in fields.items():
        if key in _RECORD_STANDARD:
            raise ValueError(f"reserved log record field: {key}")
        extra[key] = value
    return extra

File: apps/react-node/server/controller-logging.ts
/** Structured controller logging (see .cursor/skills/controller-logging/). */

import type { Request } from "express";

import { requestBody, requestHeaders, requestUrlParams } from "./request-snapshot.js";
import { currentSessionId } from "./request-context.js";
import { writeLog } from "./observability-logging.js";

type Fields = Record<string, unknown>;
type LogLevel = "INFO" | "WARN" | "ERROR" | "TRACE";

function emit(
  level: LogLevel,
  message: string,
  handler: string,
  source: string,
  fields?: Fields,
): void {
  const sessionId = currentSessionId();
  const extra = {
    source,
    controller: handler,
    ...(sessionId ? { session_id: sessionId } : {}),
    ...fields,
  };
  const fileLevel = level === "TRACE" ? "TRACE" : level;
  writeLog(fileLevel, message, extra);
  const line = JSON.stringify({ level: fileLevel, message, ...extra });
  if (level === "ERROR") {
    console.error(line);
  } else if (level === "WARN") {
    console.warn(line);
  } else {
    console.info(line);
  }
}

export function httpRequestFields(req: Request): Fields {
  return {
    headers: requestHeaders(req),
    url_params: requestUrlParams(req),
    body: requestBody(req),
  };
}

export function logReceived(
  handler: string,
  source: string,
  method: string,
  path: string,
  fields?: Fields,
): void {
  emit("INFO", `${handler} request received`, handler, source, { method, path, ...fields });
}

export function logReceivedFromRequest(
  req: Request,
  handler: string,
  source: string,
  method: string,
  path: string,
  fields?: Fields,
): void {
  logReceived(handler, source, method, path, {
    ...httpRequestFields(req),
    ...fields,
  });
}

export function logSucceeded(handler: string, source: string, fields?: Fields): void {
  emit("INFO", `${handler} succeeded`, handler, source, fields);
}

export function logWarn(
  handler: string,
  source: string,
  message: string,
  fields?: Fields,
): void {
  emit("WARN", message, handler, source, fields);
}

export function logError(
  handler: string,
  source: string,
  message: string,
  fields?: Fields,
): void {
  emit("ERROR", message, handler, source, fields);
}

export function logTrace(
  handler: string,
  source: string,
  message: string,
  fields?: Fields,
): void {
  emit("TRACE", message, handler, source, fields);
}
